docs/error_mitigation/readout_mitigation.md

```diff
@@ -0,0 +1,257 @@
+mitigated_samples_mle = mitigation(model=model, noise=noise)
+print(f"MLE based mitigation {mitigated_samples_mle}")
+```
+
+### Matrix free measurement mitigation (MTHREE)
+
+This method relies on inverting the probability distribution within an identified subspace [^3] (restricting to the space of measured bitstrings). The method is reserved for computations that exceed 20 qubits where trying to recreate the corrected probability distribution would be creating a state in a unreasonably high dimensional hilbert space. Thus the idea here is to stick to the basis states that show up in the measurement alone. Additionally one might want to include states that are k hamming distance away from it.
+
+
+```python exec="on" source="material-block" session="m3" result="json"
+
+
+import numpy as np
+from scipy.sparse import csr_matrix
+from scipy.sparse.linalg import gmres
+
+n_qubits = 10
+exact_prob = np.random.rand(2 ** (n_qubits))
+exact_prob[2 ** (n_qubits//2):]=0
+exact_prob = 0.90 * exact_prob + 0.1 * np.ones(2**n_qubits) / 2**n_qubits
+exact_prob = exact_prob / sum(exact_prob)
+np.random.shuffle(exact_prob)
+
+observed_prob = np.array(exact_prob, copy=True)
+observed_prob[exact_prob < 1 / 2 ** (n_qubits)] = 0
+
+observed_prob = observed_prob / sum(observed_prob)
+
+input_csr = csr_matrix(observed_prob, shape=(1, 2**n_qubits)).T
+print({bin(x)[2:].zfill(n_qubits):np.round(input_csr[x,0],3) for x in input_csr.nonzero()[0]})
+print(f"Filling percentage {len(input_csr.nonzero()[0])/2**n_qubits} %")
+```
+
+We have generated a probability distribution within a small subspace of bitstrings being filled. We use $csr\_matrix$ for efficient representation and computation. Now we use MTHREE to do mitigation on the probability distribution
+
+
+```python exec="on" source="material-block" session="m3" result="json"
+from scipy.stats import wasserstein_distance
+from qadence_protocols.mitigations.readout import (
+    normalized_subspace_kron,
+    mle_solve,
+    matrix_inv, 
+    tensor_rank_mult
+    )
+
+noise_matrices = []
+for t in range(n_qubits):
+    t_a, t_b = np.random.rand(2) / 8
+    K = np.array([[1 - t_a, t_a], [t_b, 1 - t_b]]).transpose()  # column sum be 1
+    noise_matrices.append(K)
+
+confusion_matrix_subspace = normalized_subspace_kron(noise_matrices, observed_prob.nonzero()[0])
+
+p_corr_mthree_gmres = gmres(confusion_matrix_subspace, input_csr.toarray())[0]
+p_corr_mthree_gmres_mle = mle_solve(p_corr_mthree_gmres)
+
+noise_matrices_inv = list(map(matrix_inv, noise_matrices))
+p_corr_inv_mle = mle_solve(tensor_rank_mult(noise_matrices_inv, observed_prob))
+
+print(wasserstein_distance(p_corr_mthree_gmres_mle, p_corr_inv_mle))
+
+```
+
+We have used wassestein_distance instead of kl_divergence as many of the bistrings have 0 probabilites. If the expected solution lies outside the space of observed bistrings, MTHREE will fail. We next look at majority voting to circument this problem when the expected output is a single bitstring.
```

Comment:
```suggestion
We have used `wassestein_distance` instead of `kl_divergence` as many of the bistrings have 0 probabilites. If the expected solution lies outside the space of observed bitstrings, `MTHREE` will fail. We next look at majority voting to circument this problem when the expected output is a single bitstring.
```